🔀 [hanse & sally] 로또 2단계 - 보너스 번호 추가

rkolx opened the pull request:
# java-lotto
마스터즈 멤버스 2022 로또 게임 프로젝트


- pair programming (hanse, sally)

#### 기능요구사항

- [x] 2등을 위해 추가 번호를 하나 더 추첨한다. 당첨 통계에 2등도 추가해야 한다.



#### 프로그래밍 요구사항

- [x] enum을 적용해 프로그래밍을 구현한다.

#### 출력 결과

<img src="https://user-images.githubusercontent.com/92699009/155632776-f8de735e-2021-4985-8104-3e79c7f7317c.PNG">

rkolx committed:
feat : Input 클래스, Main 클래스

- 보너스볼 입력 추가

rkolx committed:
refactor : 당첨번호 모음 클래스 및 클래스 명 수정

- WinningNumber 클래스 생성
- Winnings 클래스에서 WinningNumberCouter로 클래스 명 수정

rkolx committed:
feat : 보너스 번호 추가

- Lottos 클래스 : 보너스 번호 추가하여 개수 세기

sally-ksh committed:
refactor: 로또 결과 로직 리팩토링

- Main.class -> LottoPlay
- Lottos getNumberOfWinningAboveThree() -> Ranking 에게 전달
- Ranking
  - this.ranking 반환을 수정하게 위해 String으로의 변경로직을 옮김

sally-ksh committed:
refactor: 보너스볼 추가 사항 추상화

- BasicWinningNumbers
  - 기존의 로또 결과 출력 기능
- AddedWinningNumbers
  - 기존의 로또 결과에 추가되는 보너스 숫자들 전달
  - BonusWinningNumbers 구현체

sally-ksh committed:
refactor: LottoPlay 비즈니스 로직 분리

- UserLotto
 - 입력받은 사용자 요청을 통해 로또 생성, 보관, 당첨 결과와의 매칭 기능을 가진다.

sally-ksh committed:
feat: 패키지 분리

- 비즈니스 로직을 나누면서 사용자가 가진 티켓 정보와 당첨 정보를 패키지로 분리

sally-ksh committed:
refactor: 보너스볼 상속으로 변경

sally-ksh committed:
fix

rkolx committed:
refactor : 로또 담는 메서드 역할에 따라 분리

- Lottos 클래스 : getTotalLottos()에서 구매한 개수만큼 로또 번호 담는 getPurchassedLotoos() 추가 분리

rkolx committed:
docs : 2단계 요구사항에 맞게 RADME.md 수정

rkolx committed:
refacfor : gettotalLottos 메서드 분리

- Lottos 클래스 : getTotalLottos() 내 로또 번호 담는 부분 toLottos() 생성하여 분리

rkolx committed:
refactor : 동작 클래스명 수정

- LottoPlay클래스명에서 LottoApplication클래스명으로 변경

wheejuni commented on src/LottoApplication.java:
```diff
@@ -0,0 +1,31 @@
+import views.PurchasedLotto;
+
+public class LottoApplication {
+	private static LottoMachine lottoMachine = new LottoMachine();
+	private static Lottos lottos = new Lottos();
+	private static UserLotto userLotto = new UserLotto(lottoMachine, lottos);
```
이 부분을 어떻게 개선하면 좋을지 3단계에서 고민해주세요~

힌트: `LottoGame` 분리, 의존성 주입

sally-ksh replied:
으-음.... 어렵지만, 알 것도 같고... @_@고민 해서 개선되도록 노력하겠습니다.
3단계 진행하면서 오토,수동만 di로 했는데도,  보지 못한 부분이 있는 거 같아요.
알려주셔서 감사합니다.🥇

wheejuni commented on src/LottoApplication.java:
```diff
@@ -0,0 +1,31 @@
+	public static void main(String[] args) {
+		PurchasedLotto purchasedLotto = purchaseLotto();
+
+		List<List<Integer>> totalLottos = userLotto.getTickets(purchasedLotto);
```
`List<List<T>>` 와 2차원 배열과의 차이점이 무엇일까요?

rkolx replied:
2차원 배열의 차이점과 리스트의 장점을 알고 리스트를 선택해서 사용한건지에 대한걸 말씀하시는 걸까요? 어떤 의미인지 몰라 여쭙니다ㅠㅠ

wheejuni replied:
네 차이점이 없다는게 정답이었어요. ㅎㅎ
리팩토링을 부탁드린다는 의미이기도 했구요~

wheejuni commented on src/LottoApplication.java:
```diff
@@ -0,0 +1,31 @@
+		List<List<Integer>> totalLottos = userLotto.getTickets(purchasedLotto);
+		showLottos(totalLottos);
+
+		List<Integer> inputValueOfWinningNumbers = inputWinningNumbers();
+		int bonusNumber = getBonusNumber();
+
+		Ranking ranking = userLotto.getRankingFromWinningNumbers(inputValueOfWinningNumbers, bonusNumber);
```
`inputValueOfWinningNumbers`, `bonusNumber` 는 단순히 파라메터로 전달되기 위한 변수들이군요.
그렇다면 별도의 선언 없이 아래와 같이 해도 되지 않을까 싶네요.

```java
Ranking ranking = userLotto.getRankingFromWinningNumbers(inputWinningNumbers(), getBonusNumber());

sally-ksh replied:
객체타입 반환의 경우는 명시적으로 적어주는 게 안정성 등에서 좋다고 알고 있었습니다. 
사실...위 경우처럼 동일한 의미가 담긴 메서드명하고 참조변수명 두 번 고려하는게 정말 의미 없기도 했었습니다.

혹시 문제되거나 안 좋은 이유가 있다면 구체적으로 알고 싶어서요.
알려주시면 감사하겠습니다.

wheejuni commented on src/domains/users/Lotto.java:
```diff
@@ -23,7 +23,7 @@ public Lotto(List<Integer> sixNumbers) {
 	}
 
-	public int numberOfWinnings(List<Integer> winningNumbers){
+	public int countNumberOfWinnings(List<Integer> winningNumbers){
```
`stream` 사용해서 개선할 수 있을 거 같네요

wheejuni commented on src/domains/users/Lotto.java:
```diff
@@ -37,4 +37,11 @@ private int getCount(int count, Integer winningNumber) {
 	}
+
+	public boolean getBonus(int bonusNumber) {
+		if (sixNumbers.contains(bonusNumber)){
+			return true;
+		}
+		return false;
```
이렇게 바꿀 수 있지 않을까요
```suggestion
        return sixNumbers.contains(bonusNumber));

wheejuni commented on src/domains/users/LottoMachine.java:
```diff
@@ -1,4 +1,4 @@
-package domains;
+package domains.users;
```
패키지명에 복수형을 잘 쓰지 않습니다.

sally-ksh replied:
현재는 한명의 입력만 받아서 복수형 쓰지 않는게 맞고, 스프링의 경우에는 여러 사용자 정보를 다루니 복수형을 쓰는게 맞다고 생각하면 될까요?
스프링 강의에서 패키지명에 복수형으로 쓰라고 하셨어서,.. 좀 더 구체적으로 알고 정리하고자 질문 합니다.
감사합니다.

wheejuni replied:
일단 `domains` 까지는 잘 이해가 되었구요, 
`users` 가 복수형이 된 이유를 알고 싶었어요. 복수형이 되면 안된다고 생각했구요.
`user` 라는 일반적인 개념, 혹은 프로그램적 요구사항을 구현한 코드들의 집합이라 `users` 라는 건 좀 어색하다고 봤거든요. 우리가 추상화해야 할 개념 자체는 `user` 라고 표현되는게 맞다고 봐서요.
스프링 강의에서 패키지명에 대해 어떻게 다루고 있는지는 잘 모르겠네요.

sally-ksh replied:
네, 말씀해 주신 부분을 이해할 수 있을거 같아요.
저는 단순 지식이었는데, 설명 감사합니다.  프로젝트는 단수형으로 변경했습니다. :)

wheejuni commented on src/domains/users/Lottos.java:
```diff
@@ -32,12 +40,17 @@ private void purchased(ArrayList<Integer> pickedNumber){
+        List<Integer> winning = winningNumbers.getNumbers();
+
+        for (Lotto lotto : lottos) {
+            int winningCount = lotto.countNumberOfWinnings(winning);
+            if (winningCount >= MINIMUM_NUMBER_OF_WINNING) {
+                boolean checkedBonus = lotto.getBonus(winningNumbers.getBonus());
+                ranking.record(checkedBonus, winningCount);
+            }
```
`Lotto` 가 갖고 있는 당첨 번호 갯수를 물어보고, 보너스까지 꺼내와서 체크한 다음 `Ranking` 객체에 set을 해주고 있군요.
그러지 말고 `Lotto` 객체가 직접 `Ranking` 결과값을 주는 방식으로 개선해야 하지 않을까요?

sally-ksh replied:
중간에 WinningNumber로 분리하고, Ranking 반환 등 개선 하면서, 시야가 좁아지고, 좀 더 분명하게 정리하지 못 한거 같아요. 
감사합니다.

wheejuni commented on src/domains/users/Lottos.java:
```diff
@@ -14,11 +15,18 @@ public Lottos() {
         }
+    }
 
+    private List<List<Integer>> getPurchasedLottos() {
```
당첨번호 리스트를 보관할 만한 객체를 설계해서 그 객체의 리스트로 전달해주면 더 좋을텐데요?

sally-ksh replied:
넵, 맞습니다.

wheejuni commented on src/domains/users/UserLotto.java:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	public List<List<Integer>> getTickets(PurchasedLotto purchasedLotto) {
+		ArrayList<ArrayList<Integer>> tickets = lottoMachine.getTicket(purchasedLotto.getNumberOfTicket());
```
`java.util.ArrayList` 에만 존재하는 메서드나 로직에 애플리케이션이 강결합되어야 할 특별한 이유가 있을까요?
그렇지 않다면 항상 `List<T>` 를 사용하는 것이 좋습니다.
이유는 여러가지가 있는데요... https://stackoverflow.com/questions/9852831/polymorphism-why-use-list-list-new-arraylist-instead-of-arraylist-list-n 를 참고해보세요.

sally-ksh replied:
요구사항 중에 "ArrayList를 사용해서 구현한다." 가 있었는데 단순하게 반환타입 까지 써야 한다고 생각하고 작업했던거 같습니다. 
이번에 페어 하면서, 위에서 언급된 부분들도 그렇고 
체력과 집중력이 많이 떨어져서 끝까지 정리 안 한 부분들이 있었는데, 
좀더 집중해서 개선하도록 이끌어주셔서 감사합니다.

또 보다 좋은 지식 알게 된 거 같아요. 감사합니다.

wheejuni reviewed (changes requested):
수고 많으셨습니다 💯 
코멘트 참고해주세요~

sally-ksh committed:
refactor: 리뷰 반영

- Lotto 객체 내부에서 직접 Ranking 체크
- Tickets : 사용자가 구매한 로또 목록들 (WinningNumber : 당첨번호 리스트)
- ArrayList -> List 변경
- 패키지명 단수형으로 변경
- for문 stream으로 개선
- boolean 반환하는 함수이용시 불필요한 조건문 제거

sally-ksh committed:
Merge pull request #1 from sally-ksh/sally-step2

refactor: 리뷰 반영
